parsing/location.ml

```diff
@@ -225,6 +225,89 @@ let print_locs ppf locs =
   Format.pp_print_list ~pp_sep:(fun ppf () -> Format.fprintf ppf ",@ ")
     print_loc ppf locs
 
+(******************************************************************************)
+(* An interval set structure; additionally, it stores user-provided information
+   at interval boundaries.
+
+   The implementation provided here is naive and assumes the number of intervals
+   to be small, but the interface would allow for a more efficient
+   implementation if needed.
+
+   Note: the structure only stores maximal intervals (that therefore do not
+   overlap).
+*)
+
+module IMapSet : sig
+  type 'a t
+  (* bounds are included *)
+  val of_intervals : (('a * int) * ('a * int)) list -> 'a t
+
+  val mem : 'a t -> pos:int -> bool
+  val find_in : 'a t -> s:int -> e:int -> 'a option
+
+  val is_start : 'a t -> pos:int -> 'a option
+  val is_end : 'a t -> pos:int -> 'a option
+
+  val extrema : 'a t -> (('a * int) * ('a * int)) option
+end
+=
+struct
+  type 'a boundary = 'a * int
+
+  (* non overlapping intervals *)
+  type 'a t = ('a boundary * 'a boundary) list
+
+  let of_intervals intervals =
+    let pos =
+      List.map (fun ((a, x), (b, y)) ->
+        if x > y then [] else [((a, x), `S); ((b, y), `E)]
+      ) intervals
+      |> List.flatten
+      |> List.sort (fun ((_, x), k) ((_, y), k') ->
+        (* Make `S come before `E so that consecutive intervals get merged
+           together in the fold below *)
+        let kn = function `S -> 0 | `E -> 1 in
+        compare (x, kn k) (y, kn k'))
+    in
+    let nesting, acc =
+      List.fold_left (fun (nesting, acc) (a, kind) ->
+        match kind, nesting with
+        | `S, `Outside -> `Inside (a, 0), acc
+        | `S, `Inside (s, n) -> `Inside (s, n+1), acc
+        | `E, `Outside ->
+            raise (Invalid_argument "ISet.of_intervals: invalid intervals")
+        | `E, `Inside (s, 0) -> `Outside, ((s, a) :: acc)
+        | `E, `Inside (s, n) -> `Inside (s, n-1), acc
+      ) (`Outside, []) pos in
+    if nesting <> `Outside then
+      raise (Invalid_argument "ISet.of_intervals: invalid intervals");
```

Comment:
When are intervals `invalid` in this sense?

Should the exception name be changed?

Reply:
If I read the implementation correctly, I don't see how this exception can ever be raised given that the bounds only inserted when the `S is before the `E. (So we have never read more Es than Ss, and we have always read exactly at much at the end). So it could be an `assert (nesting = `Outside)`.

Reply:
Indeed, the exceptions were actually impossible cases. I turned them into `assert false`.